### PR TITLE
app bandwidth model implementation

### PR DESCRIPTION
**Pull request by ZhimaoL**

### CODE REVIEW

File: x/bandwidth/manager/manager.go [hunk 1]
```diff
@@ -208,6 +228,217 @@ func (bm BandwidthManager) DecayMaxMPS(ctx sdk.Context) sdk.Error {
...
+				ExpectedMPS:        expectedMPS,
+				LastRefilledAt:     ctx.BlockHeader().Time.Unix(),
+			}
+
+			if err := bm.storage.SetAppBandwidthInfo(ctx, app.Username, &newAppInfo); err != nil {
+				return err
+			}
+
+		} else {
+			// only update it's max and expected MPS, refill will happen when the app actually send messages.
```
Comment (ZhimaoL) [inline x/bandwidth/manager/manager.go]: should refill first since rate change

File: x/bandwidth/tick.go [hunk 2]
```diff
@@ -27,5 +27,36 @@ func EndBlocker(
 	if err := bm.UpdateMaxMPSAndEMA(ctx); err != nil {
 		panic(err)
 	}
+
```
Comment (ZhimaoL) [inline x/bandwidth/tick.go]: into bandwidth manager

File: x/bandwidth/manager/manager.go [hunk 3]
```diff
@@ -208,6 +228,217 @@ func (bm BandwidthManager) DecayMaxMPS(ctx sdk.Context) sdk.Error {
...
+	return nil
+}
+
+func (bm BandwidthManager) CheckBandwidth(ctx sdk.Context, accKey linotypes.AccountKey, fee auth.StdFee) sdk.Error {
+	if bm.dm.DoesDeveloperExist(ctx, accKey) {
+		appName, err := bm.dm.GetAffiliatingApp(ctx, accKey)
+		if err != nil {
+			return err
+		}
+		if !bm.IsAppBandwidthEnough(ctx, appName) {
```
Comment (ZhimaoL) [inline x/bandwidth/manager/manager.go]: consume u first

File: x/bandwidth/manager/manager.go [hunk 4]
```diff
@@ -208,6 +228,217 @@ func (bm BandwidthManager) DecayMaxMPS(ctx sdk.Context) sdk.Error {
...
+
+func (bm BandwidthManager) GetPunishmentCoeff(ctx sdk.Context, accKey linotypes.AccountKey) (sdk.Dec, sdk.Error) {
+	lastBlockTime, err := bm.gm.GetLastBlockTime(ctx)
+	if err != nil {
+		return sdk.NewDec(1), err
+	}
+	if lastBlockTime >= ctx.BlockHeader().Time.Unix() {
+		return sdk.NewDec(1), nil
+	}
+	pastTime := ctx.BlockHeader().Time.Unix() - lastBlockTime
```
Comment (ZhimaoL) [inline x/bandwidth/manager/manager.go]: pasttime

File: x/bandwidth/manager/manager.go [hunk 5]
```diff
@@ -208,6 +273,262 @@ func (bm BandwidthManager) DecayMaxMPS(ctx sdk.Context) sdk.Error {
...
+			if err := bm.storage.SetAppBandwidthInfo(ctx, app.Username, info); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (bm BandwidthManager) CheckBandwidth(ctx sdk.Context, accKey linotypes.AccountKey, fee auth.StdFee) sdk.Error {
+	if bm.dm.DoesDeveloperExist(ctx, accKey) {
```
Comment (xiaoxiaff) [inline x/bandwidth/manager/manager.go]: Does affiliate account return true from this call? @Stumble

Comment (Stumble) [reply x/bandwidth/manager/manager.go]: nope

**Pull request merged**